ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/StreamSelectBuilder.java

```diff
@@ -54,12 +66,43 @@ private StreamSelectBuilder() {
     final Named selectName =
         Named.as(StreamsUtil.buildOpName(queryContext));
 
-    return stream.withStream(
-        stream.getStream().transformValues(
-            () -> new KsTransformer<>(selectMapper.getTransformer(logger)),
-            selectName
-        ),
-        selection.getSchema()
-    );
+    if (!selectedKeys.get().containsAll(
+        sourceSchema.key().stream().map(Column::name).collect(ImmutableList.toImmutableList())
+    )) {
+      return stream.withStream(
+          stream.getStream().transform(
+                  () -> new KsTransformer<>(new KsqlTransformer<K, K>() {
```

Comment:
We might want to extract this into its own class before merging (eg `KsKeyTransformer`)